zerver/lib/presence.py

```diff
@@ -26,15 +26,40 @@ def get_presence_dicts_for_rows(
 
     for presence_row in all_rows:
         user_key = get_user_key(presence_row)
+
+        last_active_time = user_presence_datetime_with_date_joined_default(
+            presence_row["last_active_time"], presence_row["user_profile__date_joined"]
+        )
+        last_connected_time = user_presence_datetime_with_date_joined_default(
+            presence_row["last_connected_time"], presence_row["user_profile__date_joined"]
+        )
+
         info = get_user_presence_info(
-            presence_row["last_active_time"],
-            presence_row["last_connected_time"],
+            last_active_time,
+            last_connected_time,
         )
         user_statuses[user_key] = info
 
     return user_statuses
 
 
+def user_presence_datetime_with_date_joined_default(
+    dt: Optional[datetime.datetime], date_joined: datetime.datetime
+) -> datetime.datetime:
+    """
+    Our data models support UserPresence objects not having None values for
+    last_active_time/last_connected_time. The API however has always sent timestamps, so for backward
+    compatibility we cannot send such values through the API and need to default to a sane datetime.
+
+    This helper functions expects to take a last_active_time or last_connected_time value and the
+    date_joined of the user, which will serve as the default value if the first argument is None.
+    """
+    if dt is None:
+        return date_joined
+
+    return dt
```

Comment:
This function may be a bit pointless, but it lets us avoid having lots of annoying `if`s in the code and drives home the idea that `date_joined` is to be used as the fallback if the presence timestamp is `None`  (instead of that decision being made in multiple places separately, allowing inconsistencies in the future if this behavior is changed in one place but not the other) and provides a nice place to just have a docstring explaining what's going on.

Reply:
This lgtm though I'd maybe call it "the legacy API" somewhere, just so this comment makes sense once we've changed the main API to something better.

(Also you should generally line-wrap prose at ~70 characters, rather than the 100 that Black will do; this avoids the GitHub scrollbar above and is easier to read in other settings too).

Reply:
Tweaked the docstring to mention "legacy API" in one place and changed the docstring  to have max line length 70. Aside of that, these latest changes are still waiting for review.